Add Firecracker post (#13)
model-checking/kani-verifier-blog: _posts/2022-07-13-using-the-kani-rust-verifier-on-a-firecracker-example.md

---

In this post we'll apply the [Kani Rust Verifier](https://github.com/model-checking/kani) (or Kani for short), our open-source formal verification tool that can prove properties about Rust code, to an example from [Firecracker](https://firecracker-microvm.github.io/), an open source virtualization project for serverless applications.
We will use Kani to get a strong guarantee that Firecracker's block device is correct with respect to a simple virtio property when parsing guest requests, which may be invalid or malicious.
In this way, we show how Kani can complement Firecracker's defense in depth investments, such as fuzzing.

Firecracker is a Virtual Machine Monitor (VMM) for lightweight virtual machines (microVMs).
A virtual machine is an abstract compute environment (i.e., a set of resources such as processors, memory and I/O) that is complete enough to run an operating system (also called a guest in this context).
A Virtual Machine Monitor (VMM) is a specialized systems component responsible for managing and running virtual machines (in an analogous way to how an operating system would manage and run processes).[^footnote-terminology]

> Firecracker was designed to meet the requirements of serverless and container applications, such as AWS Lambda and AWS Fargate.

Reviewer: Add links to lambda and fargate.

The central difficulty is that devices must be exposed by Firecracker to a guest (in order for the guest to do useful work) but Firecracker cannot trust the guest to be well-behaved.
Firecracker's [design doc](https://github.com/firecracker-microvm/firecracker/blob/main/docs/design.md#threat-containment) says: _"from a security perspective, all vCPU threads [running the guest] are considered to be running malicious code as soon as they have been started; these malicious threads need to be contained"_.
This is important because bugs in code responsible for containment, such as device emulation, are potential security issues.
For example, a buffer overflow bug in Firecracker's vsock device was the root cause of a [CVE](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2019-18960) that could _"be used by a malicious guest to read from and write to a segment of the host-side Firecracker process' heap address space"_ ([source](https://github.com/firecracker-microvm/firecracker/issues/1462)).

One of the select number of devices exposed by Firecracker to the microVM is a block device.
(A block device, such as a disk, is a device that allows random access of data in fixed-size blocks.)

> In Firecracker, the block device is a piece of code that *emulates* a physical block device from the perspective of the guest but in reality is a disk image that is specified by the user when setting up the microVM.[^footnote-api]

Reviewer: ```suggestion
In Firecracker, the block device is a piece of code that *emulates* a physical block device from the perspective of the guest, but in reality is a disk image that is specified by the user when setting up the microVM.[^footnote-api]
```

The `flags` field indicates whether the buffer is read-only or write-only from the point of view of the device (a buffer cannot be both readable and writable).
Additionally, if `flags` indicates that the descriptor is "chained" then the `next` field gives the index in the descriptor table of the subsequent descriptor.
Descriptor chaining allows multiple buffers to be grouped into a logical list (also known as a *scatter/gather list*).

The virtio standard does not specify how a request should be chopped up into descriptors/buffers.
However, the Firecracker block device expects that read requests use a descriptor chain of length `3`:
  - The first describes a read-only buffer that specifies a read request type (`VIRTIO_BLK_T_IN`) and the sector to be read (multiplying the sector by the block size, which is 512 bytes for Firecracker, gives the offset).
  - The second describes a write-only buffer for the device to fill with the contents of the requested sector. The length of this buffer is the requested size.
  - The third describes a write-only buffer for the device to fill with a status byte indicating whether the request succeeded.

> A write request is similar except that the first buffer specifies a write request type (`VIRTIO_BLK_T_OUT`) and the second descriptor describes a read-only buffer for the device that contains the data to be written-to the requested sector.

Reviewer: ```suggestion
A write request is similar except that the first buffer specifies a write request type (`VIRTIO_BLK_T_OUT`) and the second descriptor describes a read-only buffer for the device that contains the data to be written to the requested sector.
```

> 2.6.4.2 Driver Requirements: Message Framing
>
> The driver MUST place any device-writable descriptor elements after any device-readable descriptor elements.
>
> Source: <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-280004>

This is a requirement on the driver under the control of the guest.
However, Firecracker cannot trust that this is the case: the requirement has to be validated.
Here's how we can turn it into a requirement on the `parse` method: it must be the case that if `parse` succeeds (i.e., returns `Ok(...)`) then the message framing requirement was met (and, also, if the requirement is not met then `parse` must fail).
Note that it is acceptable for the `parse` method to fail even if this requirement is met.
For example, a descriptor chain may be properly message framed, but specify an invalid request type.

Imagine for a moment that we had a checker that could tell us whether the sequence of `Descriptors` returned by `mem` adhered to this requirement.
Then we could write:

```rust
let result = Request::parse(/*--snip--*/);
if result.is_ok() {
    assert!(checker.virtio_2642_holds())
}
if !checker.virtio_2642_holds() {

>     assert result.is_err();

Reviewer: ```suggestion
    assert!(result.is_err());
```

Our example is available in Kani's test directory with instructions on reproducing the results of this post (see [Further Reading](#further-reading)).

### Using `kani::any::<T>()` to mock guest memory

In earlier posts ([here](https://model-checking.github.io/kani-verifier-blog/2022/05/04/announcing-the-kani-rust-verifier-project.html#enter-kani) and [here](https://model-checking.github.io/kani-verifier-blog/2022/06/01/using-the-kani-rust-verifier-on-a-rust-standard-library-cve.html#bounded-results)) we introduced Kani [`any<T>()`](https://model-checking.github.io/kani/tutorial-nondeterministic-variables.html).
This is a feature that informally generates "any `T` value".
This is not the same as a randomly chosen concrete value (as in fuzzing) but rather a *symbolic* value that represents any possible value of the appropriate type.
The key idea is that we can use `any` in a Kani harness to verify the behavior of our code with respect to all possible values of an input (rather than having to exhaustively enumerate them).

In previous posts, we've mostly used `any` in proof harnesses (the entry point for analysis analogous to a test in unit testing).
We'll go a bit further here.

> When we look at the `parse` method a good place to think about using `any` is the guest memory `mem`.

Reviewer: ```suggestion
When we look at the `parse` method, a good place to think about using `any` is the guest memory `mem`.
```

```rust
    fn read_obj<T>(&self, addr: GuestAddress) -> Result<T, Error>
    where
        T: ByteValued + kani::Invariant + ReadObjChecks<T>,
    {
        if kani::any() {
            let val = kani::any::<T>();
            T::check_on_read_val(&self, &val);
            Ok(val)
        } else {
            Err(kani::any::<Error>())
        }
    }
```

Now let's address the `ReadObjChecks` trait bound that we added, which allows us to call `T::check_on_read_val` where we pass in the generated symbolic value.
This allows us to attach our checker to calls of `read_obj` where `T = Descriptor`.
In this way, we can update our state machine.

```rust
trait ReadObjChecks<T> {
    type CheckerType;
    fn check_on_read_val(mem: &GuestMemoryMmap, read_val: &T);
}

impl ReadObjChecks<Descriptor> for Descriptor {
    type CheckerType = DescriptorPermissionChecker;
    fn check_on_read_val(mem: &GuestMemoryMmap, read_val: &Descriptor) {
        let current_permission = DescriptorPermission::from_flags(read_val.flags);
        mem.permission_checker
            .borrow_mut()
            .update(current_permission);
    }
}
```

### Proof harness

> Putting this altogether, we can write a harness as follows.

Reviewer: ```suggestion
Putting this all together, we can write a harness as follows.
```

At a high-level, we generate symbolic inputs for the `parse` and then assert our property of interest.

> The only nit is that we must generate an initial valid `DescriptorChain` (not all `kani::any<DescriptorChain>()` values are valid).

Reviewer: ```suggestion
The only nit is that we must generate an initial valid `DescriptorChain` (not all `kani::any::<DescriptorChain>()` values are valid).
```

The way we chose to do this is to call `DescriptorChain::checked_new` with symbolic inputs and proceed if this returns a valid `DescriptorChain`.

```rust
#[cfg(kani)]
mod verification {
    use super::*;

>     // ANCHOR: stretch_kani

Reviewer: Remove this line.

    #[kani::proof]
    pub fn requirement_2642() {
        let mem = GuestMemoryMmap::new();
        let desc_table: GuestAddress = kani::any();
        let queue_size: u16 = kani::any();
        let index: u16 = kani::any();
        let desc = DescriptorChain::checked_new(&mem, desc_table, queue_size, index);
        match desc {
            Some(x) => {
                let req = Request::parse(&x, &mem, kani::any::<u64>());
                if req.is_ok() {
                    assert!(mem.permission_checker.borrow().virtio_2642_holds());
                }
                if !(mem.permission_checker.borrow().virtio_2642_holds()) {
                    assert!(req.is_err());
                }
            }
            None => {}
        };
    }
}
```

Passing the harness to Kani results in `VERIFICATION:- SUCCESSFUL` in a few seconds.

> As sanity check, if we insert an issue such as forgetting a validity check to ensure the third buffer is read-only then Kani reports an error (since in this case it is possible to fail virtio requirement 2.6.4.2 but return a valid `Request`).

Reviewer: ```suggestion
As sanity check, if we insert an issue such as forgetting a validity check to ensure the third buffer is read-only, then Kani reports an error (since in this case it is possible to fail virtio requirement 2.6.4.2 but return a valid `Request`).
```

Our example is available in Kani's test directory with instructions on reproducing these results.

But what does this result mean?
A way to approximate the power of this approach is to think about the number of test cases we would need to write to get the same level of assurance.
Ignoring error cases, which are also covered, our harness using our mock guest memory enables Kani to explore the behavior of `parse` when it reads up to `3` symbolic descriptors.
These are each 16-byte values.
The `parse` function also reads the first buffer to determine the request type and sector.
This is a 16-byte value.
In this case that's `2^16 * 2^16 * 2^16 * 2^16 = 2^64` test cases.
Kani allows us to definitively check a property of interest (if virtio requirement 2.6.4.2 does not hold then `parse` will fail) across all of these cases.

## Summary

In this post, we introduced Firecracker and focused our attention on its virtio block device implementation.
We encoded a finite-state machine to check a simple virtio property.

> Using Kani's verification features, in particular, `kani::any::<T>()` we built a mock guest memory that uses symbolic values to model reads (of any value).

Reviewer: ```suggestion
Using Kani's verification features, in particular, `kani::any::<T>()`, we built a mock guest memory that uses symbolic values to model reads (of any value).
```
